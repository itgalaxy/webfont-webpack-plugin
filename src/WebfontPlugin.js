import path from "path";
import Watchpack from "watchpack";
import fs from "fs-extra";
import nodify from "nodeify";
import webfont from "webfont";

export default class WebfontPlugin {
  constructor(options = {}) {
    if (!options.files) {
      throw new Error("Require `files` options");
    }

    if (!options.dest) {
      throw new Error("Require `dest` options");
    }

    this.options = Object.assign(
      {
        bail: null
      },
      options
    );
    this.pluginName = "WebfontPlugin";
    this.firstRun = true;
    this.watching = null;
    this.watcher = null;
    this.needRegenerate = true;
  }

  apply(compiler) {
    this.fileDependencies = [];
    this.contextDependencies = [];

    if (typeof this.options.bail !== "boolean" && compiler.options.bail) {
      this.options.bail = compiler.options.bail;
    }

    const hookFn = callback =>
      this.generate(error =>
        callback(error && this.options.bail ? error : null)
      );

    const beforeRunFn = (compilation, callback) => hookFn(callback);
    const watchRunFn = (watching, callback) => {
      this.watching = watching;

      if (this.firstRun) {
        return hookFn(() => {
          this.firstRun = false;
          this.needRegenerate = false;

          return callback();
        });
      }

      return callback();
    };
    const doneFn = () => {
      if (this.watching && !this.watching.closed) {
        const oldWatcher = this.watcher;

        this.watcher = new Watchpack(
          this.watching.watchFileSystem
            ? this.watching.watchFileSystem.watcherOptions
            : {}
        );
        this.watcher.watch(
          this.fileDependencies,
          this.contextDependencies,
          Date.now()
        );
        this.watcher.once("change", () => {
          if (!this.needRegenerate) {
            this.needRegenerate = true;
            hookFn(() => {
              this.needRegenerate = false;
            });
          }
        });

        if (oldWatcher) {
          oldWatcher.close();
        }
      }
    };
    const watchCloseFn = () => {
      if (this.watcher) {
        this.watcher.close();
      }
    };

    if (compiler.hooks) {
      compiler.hooks.beforeRun.tapAsync(this.pluginName, beforeRunFn);
      compiler.hooks.watchRun.tapAsync(this.pluginName, watchRunFn);
      compiler.hooks.done.tap(this.pluginName, doneFn);
      compiler.hooks.watchClose.tap(this.pluginName, watchCloseFn);
    } else {
      compiler.plugin("before-run", beforeRunFn);
      compiler.plugin("watch-run", watchRunFn);
      compiler.plugin("done", doneFn);
      compiler.plugin("watch-close", watchCloseFn);
    }
  }

  generate(callback) {
    const { options } = this;

    nodify(
      webfont(options).then(result => {
        const { fontName, template } = result.config;
        const dest = path.resolve(this.options.dest);

        let destTemplate = dest;

        if (result.template) {
          if (this.options.destTemplate) {
            destTemplate = path.resolve(this.options.destTemplate);
          }

          if (result.usedBuildInTemplate) {
            destTemplate = path.join(destTemplate, `${fontName}.${template}`);
          } else {
            destTemplate = path.join(
              destTemplate,
              path.basename(template).replace(".njk", "")
            );
          }

          if (result.config.config) {
            const configFilePath = result.config.config;

            if (!this.fileDependencies.includes(configFilePath)) {
              this.fileDependencies.push(configFilePath);
            }
          }

          if (!result.usedBuildInTemplate) {
            const templateFilePath = result.config.template;

            if (!this.fileDependencies.includes(templateFilePath)) {
              this.fileDependencies.push(templateFilePath);
            }
          }

          if (result.glyphsData) {
            result.glyphsData.forEach(glyphData => {
              const { srcPath } = glyphData;
              const srcDirname = path.dirname(srcPath);

              if (!this.contextDependencies.includes(srcDirname)) {
                this.contextDependencies.push(srcDirname);
              }
            });
          }

          if (result.fontsData) {
            result.fontsData.forEach(fontData => {
              const { srcPath } = fontData;
              const srcDirname = path.dirname(srcPath);

              if (!this.contextDependencies.includes(srcDirname)) {
                this.contextDependencies.push(srcDirname);
              }
            });
          }

        }

        return Promise.all(
          Object.keys(result).map(type => {
            if (
              type === "config" ||
              type === "usedBuildInTemplate" ||
              type === "glyphsData" ||
              type === "fontsData"
            ) {
              return Promise.resolve();
            }

            let content = result[type];

            // After adding TTF mode multiple files can be returned.
            //
            if (!Array.isArray(content)) {

              if (type !== "template") {
                content = [{name: path.resolve(dest, `${fontName}.${type}`), buffer: content}];
              } else {
                content = [{name: destTemplate, buffer: content}];
              }

            } else {

              content.forEach(file => {
                file.name = path.resolve(dest, path.basename(file.name));
              })

            }


            return Promise.all(content.map(file => {
              return fs.outputFile(file.name, file.buffer);
            }));

          })
        );
      }),
      error => callback(error)
    );
  }
}
