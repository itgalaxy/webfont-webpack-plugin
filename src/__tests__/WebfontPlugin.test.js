import path from "path";
import del from "del";
import fs from "fs-extra";
import webpack from "webpack";
import WebfontPlugin from "../WebfontPlugin";
import webpackConfigBase from "./fixtures/config-base";

const fixtures = path.resolve(__dirname, "fixtures");
const cssDir = path.join(fixtures, "css");

const pluginBaseConfig = {
  dest: path.resolve(cssDir, "../css"),
  destTemplate: path.resolve(cssDir, "../css"),
  files: path.join(fixtures, "svg-icons/**/*.svg"),
  template: "css",
  templateFontPath: "./"
};

/* eslint-disable no-sync */

describe("webfontPlugin", () => {
  beforeAll(() =>
    del([
      path.resolve(fixtures, "build"),
      path.resolve(fixtures, "css/*.{svg,ttf,eot,woff,woff2,css}")
    ])
  );

  afterEach(() =>
    del([
      path.resolve(fixtures, "build"),
      path.resolve(fixtures, "css/*.{svg,ttf,eot,woff,woff2,css}")
    ])
  );

  it("should export `WebfontPlugin` as a class", () => {
    expect(typeof WebfontPlugin === "function").toBe(true);
  });

  it("should throw error if not passed `files`", () => {
    expect(() => new WebfontPlugin()).toThrow("Require `files` options");
  });

  it("should throw error if not passed `dest`", () => {
    expect(
      () =>
        new WebfontPlugin({
          files: "**!/!*.svg"
        })
    ).toThrow("Require `dest` options");
  });

  it("should generate fonts and build-in template", done => {
    const options = Object.assign({}, pluginBaseConfig);

    webpackConfigBase.plugins = [new WebfontPlugin(options)];

    webpack(webpackConfigBase, (error, stats) => {
      if (error) {
        throw error;
      }

      expect(stats.compilation.warnings).toMatchSnapshot("warnings");
      expect(stats.compilation.errors).toMatchSnapshot("errors");

      expect(fs.existsSync(path.join(cssDir, "webfont.css"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.eot"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.svg"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.ttf"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.woff"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.woff2"))).toBe(true);

      return done();
    });
  });

  it("should generate fonts and external template", done => {
    const options = Object.assign({}, pluginBaseConfig, {
      template: path.resolve(fixtures, "templates/webfont.css.njk")
    });

    webpackConfigBase.plugins = [new WebfontPlugin(options)];

    webpack(webpackConfigBase, (error, stats) => {
      if (error) {
        throw error;
      }

      expect(stats.compilation.warnings).toMatchSnapshot();
      expect(stats.compilation.errors).toMatchSnapshot();

      expect(fs.existsSync(path.join(cssDir, "webfont.css"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.eot"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.svg"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.ttf"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.woff"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.woff2"))).toBe(true);

      expect(Object.keys(stats.compilation.assets)).toMatchSnapshot();

      return done();
    });
  });

  it("should generate fonts and external config", done => {
    const options = Object.assign({}, pluginBaseConfig, {
      config: path.resolve(fixtures, "config/.webfontrc")
    });

    webpackConfigBase.plugins = [new WebfontPlugin(options)];

    webpack(webpackConfigBase, (error, stats) => {
      if (error) {
        throw error;
      }

      expect(stats.compilation.warnings).toMatchSnapshot();
      expect(stats.compilation.errors).toMatchSnapshot();

      expect(fs.existsSync(path.join(cssDir, "webfont.css"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.eot"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.svg"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.ttf"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.woff"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.woff2"))).toBe(true);

      return done();
    });
  });

  it("should regenerate fonts and build-in template in watch mode", done => {
    const options = Object.assign({}, pluginBaseConfig);

    webpackConfigBase.plugins = [new WebfontPlugin(options)];

    const compiler = webpack(webpackConfigBase);
    let countCompilation = 0;

    const watching = compiler.watch({}, (error, stats) => {
      if (error) {
        throw error;
      }

      expect(stats.compilation.warnings).toMatchSnapshot(
        `warning in compilation ${countCompilation}`
      );
      expect(stats.compilation.errors).toMatchSnapshot(
        `errors in compilation ${countCompilation}`
      );

      if (countCompilation === 0) {
        watching.close();
      }

      countCompilation++;

      expect(fs.existsSync(path.join(cssDir, "webfont.css"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.eot"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.svg"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.ttf"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.woff"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.woff2"))).toBe(true);

      return done();
    });

    Promise.resolve()
      .then(() => fs.readFile(path.join(fixtures, "svg-icons/avatar.svg")))
      .then(content =>
        fs.writeFile(path.join(fixtures, "svg-icons/avatar.svg"), content)
      )
      .catch(error => {
        throw error;
      });
  });

  it("should regenerate fonts and external template in watch mode", done => {
    const options = Object.assign({}, pluginBaseConfig, {
      template: path.resolve(fixtures, "templates/webfont.css.njk")
    });

    webpackConfigBase.plugins = [new WebfontPlugin(options)];

    const compiler = webpack(webpackConfigBase);
    let countCompilation = 0;

    const watching = compiler.watch({}, (error, stats) => {
      if (error) {
        throw error;
      }

      expect(stats.compilation.warnings).toMatchSnapshot(
        `warning in compilation ${countCompilation}`
      );
      expect(stats.compilation.errors).toMatchSnapshot(
        `errors in compilation ${countCompilation}`
      );

      if (countCompilation === 0) {
        watching.close();
      }

      countCompilation++;

      expect(fs.existsSync(path.join(cssDir, "webfont.css"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.eot"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.svg"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.ttf"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.woff"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.woff2"))).toBe(true);

      return done();
    });

    Promise.resolve()
      .then(() =>
        fs.readFile(path.resolve(fixtures, "templates/webfont.css.njk"))
      )
      .then(content =>
        fs.writeFile(
          path.resolve(fixtures, "templates/webfont.css.njk"),
          content
        )
      )
      .catch(error => {
        throw error;
      });
  });

  it("should generate and regenerate fonts and external config in watch mode", done => {
    const options = Object.assign({}, pluginBaseConfig, {
      config: path.resolve(fixtures, ".webfontrc")
    });

    webpackConfigBase.plugins = [new WebfontPlugin(options)];

    const compiler = webpack(webpackConfigBase);
    let countCompilation = 0;

    const watching = compiler.watch({}, (error, stats) => {
      if (error) {
        throw error;
      }

      expect(stats.compilation.warnings).toMatchSnapshot(
        `warning in compilation ${countCompilation}`
      );
      expect(stats.compilation.errors).toMatchSnapshot(
        `errors in compilation ${countCompilation}`
      );

      if (countCompilation === 0) {
        watching.close();
      }

      countCompilation++;

      expect(fs.existsSync(path.join(cssDir, "webfont.css"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.eot"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.svg"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.ttf"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.woff"))).toBe(true);
      expect(fs.existsSync(path.join(cssDir, "webfont.woff2"))).toBe(true);

      return done();
    });

    Promise.resolve()
      .then(() => fs.readFile(path.resolve(fixtures, ".webfontrc")))
      .then(content =>
        fs.writeFile(path.resolve(fixtures, ".webfontrc"), content)
      )
      .catch(error => {
        throw error;
      });
  });

  it("should has errors with default `bail` value", done => {
    const options = Object.assign({}, pluginBaseConfig, {
      files: [
        path.resolve(__dirname, "fixtures/svg-icons/**/*.svg"),
        path.resolve(__dirname, "fixtures/broken-svg-icons/**/*.svg")
      ]
    });

    webpackConfigBase.plugins = [new WebfontPlugin(options)];

    webpack(webpackConfigBase, (error, stats) => {
      if (error) {
        throw error;
      }

      expect(stats.compilation.warnings).toMatchSnapshot();
      expect(stats.compilation.errors).toHaveLength(1);

      expect(Object.keys(stats.compilation.assets)).toMatchSnapshot();

      done();
    });
  });
});
