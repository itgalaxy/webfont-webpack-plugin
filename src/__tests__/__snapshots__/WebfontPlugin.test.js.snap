// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webfontPlugin should generate and regenerate fonts and external config in watch mode: errors in compilation 0 1`] = `Array []`;

exports[`webfontPlugin should generate and regenerate fonts and external config in watch mode: warning in compilation 0 1`] = `Array []`;

exports[`webfontPlugin should generate fonts and build-in template: errors 1`] = `Array []`;

exports[`webfontPlugin should generate fonts and build-in template: warnings 1`] = `Array []`;

exports[`webfontPlugin should generate fonts and external config 1`] = `Array []`;

exports[`webfontPlugin should generate fonts and external config 2`] = `Array []`;

exports[`webfontPlugin should generate fonts and external template 1`] = `Array []`;

exports[`webfontPlugin should generate fonts and external template 2`] = `Array []`;

exports[`webfontPlugin should generate fonts and external template 3`] = `
Array [
  "bundle.js",
]
`;

exports[`webfontPlugin should has errors with default \`bail\` value 1`] = `Array []`;

exports[`webfontPlugin should has errors with default \`bail\` value 2`] = `
Array [
  "bundle.js",
]
`;

exports[`webfontPlugin should regenerate fonts and build-in template in watch mode: errors in compilation 0 1`] = `Array []`;

exports[`webfontPlugin should regenerate fonts and build-in template in watch mode: warning in compilation 0 1`] = `Array []`;

exports[`webfontPlugin should regenerate fonts and external template in watch mode: errors in compilation 0 1`] = `Array []`;

exports[`webfontPlugin should regenerate fonts and external template in watch mode: warning in compilation 0 1`] = `Array []`;
